### PR TITLE
Fix constant folding of integers in RuntimeBinder.

### LINKED ISSUE #15464: Microsoft.CSharp.RuntimeBinder constant folding incorrect in several ways
Given:

```C#
var x = CSharpArgumentInfo.Create(CSharpArgumentInfoFlags.Constant, null);
var y = CSharpArgumentInfo.Create(CSharpArgumentInfoFlags.Constant, null);
var binder = Microsoft.CSharp.RuntimeBinder.Binder.BinaryOperation(CSharpBinderFlags.None, ExpressionType.Divide, GetType(), new[] {x, y});
var callsite = CallSite<Func<CallSite, object, object, object>>.Create(binder);
callsite.Target(callsite, 8, 0);
```

We would expect the last line to throw a `RuntimeBinderException` due to the divide by constant zero.

Instead it throws `OverflowException` (in debug it throws an assertion failure). Of the other integer types, `uint`, `long` and `ulong` will all throw a `DivideByZeroException` while all the others throw `OverflowException`.

### PR DESCRIPTION
**Pull request by JonHanna**
Rewrite the constant folding done in RuntimeBinder to fix several bugs.

Fixes #15464

### CODE REVIEW

**Comment by JonHanna** [comment]
Test Innerloop OSX Debug Build and Test please.
(Timeout)

**Comment by JonHanna** [comment]
Test Innerloop Windows_NT Release Build and Test please

**Comment by JonHanna** [comment]
Test Innerloop Windows_NT Release Build and Test please.
(Timeout)

**Comment by danmosemsft** [comment]
@JonHanna @VSadov @OmarTawfik

**Comment by karelz** [comment]
@VSadov @OmarTawfik ping?

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Operators.cs [hunk 1]
```diff
@@ -2587,34 +2588,45 @@ private bool GetBinopKindAndFlags(ExpressionKind ek, out BinOpKind pBinopKind, o
             return true;
         }
 
-        private static bool isDivByZero(ExpressionKind kind, EXPR op2)
+        private static bool IsDivByZero(ExpressionKind kind, EXPR op2, PredefinedType ptOp)
```
Comment (VSadov) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Operators.cs]: ptOp seems unused

Comment (JonHanna) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Operators.cs]: Now. Why did I add that?

Comment (JonHanna) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Operators.cs]: Ah. I had been considering the case of floating-point values, which of course shouldn't throw, but when seeing that it is never hit with those I removed the guard against it but not the parameter.

Updated.

**Comment by JonHanna** [comment]
Closing to remove constant folding instead, as per #15467

**Pull request closed without merging**